saleor/order/utils.py

```diff
@@ -10,33 +12,39 @@
 
 
 def check_order_status(func):
-    """Prevent execution of decorated function if order is fully paid.
+    """Check if order meets preconditions of payment process.
 
-    Instead redirects to order details page.
+    Order can not have draft status or be fully paid. Billing address
+    must be provided.
+    If not, redirect to order details page.
     """
     # pylint: disable=cyclic-import
     from .models import Order
 
     @wraps(func)
     def decorator(*args, **kwargs):
         token = kwargs.pop('token')
-        order = get_object_or_404(Order, token=token)
-        if order.is_fully_paid():
+        order = get_object_or_404(Order.objects.confirmed(), token=token)
+        if not order.billing_address or order.is_fully_paid():
             return redirect('order:details', token=order.token)
         kwargs['order'] = order
         return func(*args, **kwargs)
 
     return decorator
 
 
-def recalculate_order(order):
+def recalculate_order(order, discount_amount=None):
     """Recalculate and assign total price of order.
 
     Total price is a sum of items in order and order shipping price minus
     discount amount.
     """
     prices = [line.get_total() for line in order]
     total = sum(prices, order.shipping_price)
+    # discount amount can't be greater than order total
+    order.discount_amount = min(
+        discount_amount or order.discount_amount or
```

Comment:
I'd assign it to some sort of variable and then pass it to `min()` function, it would be more readable